src/Pos/Wallet/Web/Server/Sockets.hs

```diff
@@ -91,5 +101,54 @@ type WebWalletSockets m = (MonadWalletWebSockets m, MonadIO m)
 runWalletWS :: ConnectionsVar -> WalletWebSockets m a -> m a
 runWalletWS = flip Ether.runReaderT
 
-notify :: WebWalletSockets m => NotifyEvent -> m ()
-notify msg = getWalletWebSockets >>= flip sendWS msg
+notifyAll :: WebWalletSockets m => NotifyEvent -> m ()
+notifyAll msg = do
+  var <- getWalletWebSockets
+  conns <- readTVarIO var
+  for_ (listConnections conns) $ flip sendWS msg
+
+------------------------------------
+-- Implementation of ConnectionMap
+------------------------------------
+
+data IntMapWithUnusedKey v = IntMapWithUnusedKey
+    { unusedKey :: Int
+    , mapping   :: IntMap v
+    }
+
+instance Default (IntMapWithUnusedKey v) where
+  def = IntMapWithUnusedKey 0 IM.empty
+
+type ConnectionKey = IM.Key
+type ConnectionMap = IntMapWithUnusedKey WSConnection
+
+registerConnection :: WSConnection -> ConnectionsVar -> STM ConnectionKey
+registerConnection conn var = do
+    conns <- readTVar var
+    let (key, newConns) = registerConnectionPure conn conns
+    writeTVar var newConns
+    pure key
+
+deregisterConnection :: ConnectionKey -> ConnectionsVar -> STM (Maybe WSConnection)
+deregisterConnection key var = do
+    conns <- readTVar var
+    case deregisterConnectionPure key conns of
+        Nothing -> pure Nothing
+        Just (conn, newConns) -> do
+            writeTVar var newConns
+            pure (Just conn)
+
+registerConnectionPure :: WSConnection -> ConnectionMap -> (ConnectionKey, ConnectionMap)
+registerConnectionPure conn conns =
+  let newKey = unusedKey conns + 1 in
+  (newKey, IntMapWithUnusedKey newKey (IM.insert newKey conn $ mapping conns))
+
+deregisterConnectionPure :: ConnectionKey -> ConnectionMap -> Maybe (WSConnection, ConnectionMap)
+deregisterConnectionPure key conns =
+  case IM.lookup key (mapping conns) of
+    Nothing -> Nothing
+    Just conn -> Just (conn,
+      IntMapWithUnusedKey (unusedKey conns) (IM.delete key $ mapping conns))
```

Comment:
I meant only that `\n` is at such place that it's hard to read. 
> Rewritten using the Maybe monad
But that's cool